### ISSUE TITLE
theano.tensor.extra_ops.cumprod gradient handle zero incorrectly

### ISSUE THREAD

**Issue by khaotik** [Theano/Theano#5197]
I encountered this bug when trying to implement DeepMind's DNC model, here's a snippet to reproduce:

```python3
import theano as th
import theano.tensor as T

print(th.__version__)

x = T.fvector()
y = T.extra_ops.cumprod(x)
dx = th.grad(T.sum(y*y), x)
fn = th.function([x], dx)

print(fn([1.,2.])
print(fn([1.,0.])
print(dx)
```
Outputs:
```
0.9.0dev2.dev-c1dfa69ce084fc7be0fc18ccb79f3d44e9129f73
array([ 10.,   4.], dtype=float32)
array([  2.,  nan], dtype=float32)
Elemwise{true_div,no_inplace}.0
```
Note the `nan`, which is supposed to be `0.`

Turning optimization on/off, switching between GPU/CPU does not help, I  believe it's a bug. The cause is `cumprod` being converted into division op, which gives division by zero, producing `nan`.

**Comment by lamblin** [comment]
Right, the gradient is doing a cumprod from the end, that includes the current index, but then divides that by the input.
I think we can fix that by shifting that cumprod by 1 to exclude the current input.
This would still fail if there are multiple zeros, though.

**Comment by khaotik** [comment]
I do have an O(n) algorithm to fix this problem. (my comment on [this](https://github.com/tensorflow/tensorflow/issues/3862)) thread) I don't have a working implementation yet.

**Comment by lamblin** [comment]
Great, that would indeed be a better way (I did not manage to get rid of the division either, I think there could be a way in O(l log n) for prod, not sure about cumprod yet).
